File: 1주차/3일차/조성혁.java
package ps_baekjoon;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
	
	static int n, m;
	static int[] arr;
	
	// 그룹 만들기
	// 모든 그룹의 합이 최소 x 이상일 때 그릅의 수가 m 이상인가 ?
	static boolean calc(int x) {
		int cnt = 0; // 생성된 그룹의 수
		int sum = 0; // 그룹 합
		
		for (int i = 0; i < n; i++) {
			sum += arr[i];
			
			if (sum >= x) {
				sum = 0;
				cnt++;
			}
		}
		
		return cnt >= m;
	}

	public static void main(String[] args) throws IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		
		String line = br.readLine();
		StringTokenizer st = new StringTokenizer(line);
		
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		
		arr = new int[n];
		line = br.readLine();
		st = new StringTokenizer(line);
		for (int i = 0; i < n; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		}
		
		int ans = 0;
		int left = 0;
		int right = Arrays.stream(arr).sum();
		while (left <= right) {
			int mid = (left + right) / 2;
						
			if (calc(mid)) {
				ans = Math.max(ans, mid);
				left = mid + 1;
			}
			else {
				right = mid - 1;
			}
		}
		
		sb.append(ans);
		System.out.println(sb);
	}
}

File: 4주차/16일차/조성혁.cpp
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

string str;
int K[3000001]; // K[i]: i까지의 K의 등장 횟수
int R[3000001];

int main() {
    // 문자열을 순회하며 K, R의 등장 횟수의 누적 합을 구한다.
    // 구간 [i, j]를 R의 영역이라고 간주할 때(K 섞일 수 있으나, 고려 X),
    // min(K[i-1], K[N]-K[j]) * 2 + R[j] - R[i-1]
    cin>>str;
    for (int i=1;i<=str.size();i++) {
        if (str[i-1]=='K') {
            K[i]=K[i-1]+1;
            R[i]=R[i-1];
        }
        else {
            K[i]=K[i-1];
            R[i]=R[i-1]+1;
        }
    }

    int l=1;
    int r=str.size();
    int ans=0;
    while (l<=r) {
        if (l==r&&str[l-1]=='K') break;

        int left=K[l-1];
        int right=K[str.size()]-K[r];
        ans=max(ans,min(left,right)*2+R[r]-R[l-1]);

        // 왼쪽 K가 더 적었다면 다음 K로 이동
        if (left<right) {
            l++;
            while (l<str.size()&&str[l-1]!='R') l++;
        }
        else {
            r--;
            while (r>=0&&str[r-1]!='R') r--;
        }
    }

    cout<<ans;
}
